### ISSUE TITLE
LLDB does not support armvXl, armvXb, armvXlhf, armvXbhf, armvXhf, aarch64l and aarch64b

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#39488]
|  |  |
| --- | --- |
| Bugzilla Link | [40141](https://llvm.org/bz40141) |
| Version | unspecified |
| OS | All |
| Attachments | [Patch for 5.0.2](https://user-images.githubusercontent.com/60944935/143758499-2e43304c-7a74-40cc-8ca5-e1793cfac2d3.gz), [Patch for 6.0.1](https://user-images.githubusercontent.com/60944935/143758500-6f103e1a-a859-4cdd-abf7-6affd6a17c76.gz) |
| Reporter | LLVM Bugzilla Contributor |

## Extended Description 

I tried to debug an application on armv7l-linux but it did not work. I applied the below patch, but the result was the same.
https://github.com/hpux735/swift-lldb/commit/e3e023f54c69512a77a87dfc328b3a8e75a9e185

The behavior of current LLDB + above patch:
  I set a breakpoint at line 6, but not worked after line 16.
  LLDB changed arch arm-*-linux-eabihf to armv7l-unknown-unknown at line 8-12.
  I was not able to set breakpoint at line 13.

 1	(lldb) file ~/a.out
 2	Current executable set to '~/a.out' (arm).
 3	(lldb) target list
 4	Current targets:
 5	* target #&#8203;0: /home/common/a.out ( arch=arm-*-linux-eabihf, platform=host )
 6	(lldb) b main
 7	Breakpoint 1: where = a.out`main, address = 0x000103f4
 8	(lldb) process launch --stop-at-entry
 9	Process 5659 launched: '/home/common/a.out' (arm)
10	(lldb) target list
11	Current targets:
12	* target #&#8203;0: <none> ( arch=armv7l-unknown-unknown, platform=host, pid=5659, state=stopped )
13	(lldb) b main
14	Breakpoint 2: no locations (pending).
15	WARNING:  Unable to resolve breakpoint to any actual locations.
16	(lldb) c
17	Process 5659 resuming
18	hello
19	Process 5659 exited with status = 0 (0x00000000) 
20	(lldb) 

I am using Digilent Zybo, but same problem will probably occur in RaspberryPi.

I think there is a bug in ArchSpec::MergeFrom.
LLDB loses vendor and OS name by calling ArchSpec::CoreUpdated(true) to set SubArch.
I think this problem is caused by the fact that SubArch can only be set by the constructor of llvm::Triple.
LLDB worked fine by saving those information.

I judged that the FindCoreDefinition should be modified to support cores other than armv7 and little-endian, so the above patch is not good.
I have created a patch. Are there any arm users who can try this?
I used 5.0.2 and 6.0.1 because libcxxabi of 7.0.0 failed in build on arm.

**Comment by llvmbot** [comment]
Hi!

This issue may be a good introductory issue for people new to working on LLVM. If you would like to work on this issue, your first steps are:

  1) Assign the issue to you.
  2) Fix the issue locally.
  3) [Run the test suite](https://llvm.org/docs/TestingGuide.html#unit-and-regression-tests) locally.
    3.1) Remember that the subdirectories under `test/` create fine-grained testing targets, so you can
         e.g. use `make check-clang-ast` to only run Clang's AST tests.
  4) Create a `git` commit
  5) Run [`git clang-format HEAD~1`](https://clang.llvm.org/docs/ClangFormat.html#git-integration) to format your changes.
  6) Submit the patch to [Phabricator](https://reviews.llvm.org/).
    6.1) Detailed instructions can be found [here](https://llvm.org/docs/Phabricator.html#requesting-a-review-via-the-web-interface)

For more instructions on how to submit a patch to LLVM, see our [documentation](https://llvm.org/docs/Contributing.html).

If you have any further questions about this issue, don't hesitate to ask via a comment on this Github issue.

@llvm/issue-subscribers-good-first-issue

**Comment by DavidSpickett** [comment]
From: 
https://github.com/llvm/llvm-project/blob/7f9d348eb2a54a2dd07ad9e029baef30d9d9b737/lldb/source/Utility/ArchSpec.cpp#L43

armvXl (used by Linux)
* armv7l supported
* armv8l supported

armvXb
* I found no use of such a target in llvm (There is Armv8B but B means "Base" not big endian as I assume this is supposed to?).

armvXlhf
armvXbhf
armvXhf
* I found no use of such a target in llvm. I think it is core type armvX and ABI hf as in hard float.

aarch64l
aarch64b
* I found no use of such a target in llvm. I assume l is little endian and b is big endian. 99.9% of lldb use will be on systems that are implicitly "l". We do support big endian internally but I don't know of anyone using it.

So I am going to close this issue.

If we add any of these in the future I want them added one by one with details of specific devices / distributions that actually produce them. That function does not need more guess work in it, Arm doesn't need more ways to refer to the same things if we can avoid it.
